File: src/main/java/com/hileco/cortex/primitives/heavy/BackedVmStack.java
package com.hileco.cortex.primitives.heavy;

import com.hileco.vm.primitives.VmStack;

import java.util.ArrayList;
import java.util.List;

public class BackedVmStack<T> implements VmStack<T, BackedVmStack<T>> {
    private List<T> backingList;

    public BackedVmStack() {
        this.backingList = new ArrayList<>();
    }

    private BackedVmStack(List<T> backingList) {
        this.backingList = new ArrayList<>(backingList);
    }

    public void close() {
        backingList.clear();
    }

    public BackedVmStack<T> copy() {
        return new BackedVmStack<>(backingList);
    }

    public T peek() {
        return backingList.get(backingList.size() - 1);
    }

    public void push(T value) {
        backingList.add(value);
    }

    public T pop() {
        return backingList.remove(backingList.size() - 1);
    }

    public int size() {
        return backingList.size();
    }
}

File: cortex-collections/src/main/kotlin/com/hileco/cortex/collections/backed/BackedVmStack.kt
package com.hileco.cortex.collections.backed

import com.hileco.cortex.collections.base.BaseVmStack
import java.util.*

class BackedVmStack<T> : BaseVmStack<T> {
    private var backingList: MutableList<T>

    constructor() {
        this.backingList = ArrayList()
    }

    private constructor(backingList: List<T>) {
        this.backingList = ArrayList(backingList)
    }

    override fun get(index: Int): T {
        return backingList[index]
    }

    override fun set(index: Int, value: T) {
        backingList[index] = value
    }

    override fun clear() {
        backingList.clear()
    }

    override fun close() {}

    override fun copy(): BackedVmStack<T> {
        return BackedVmStack(backingList)
    }

    override fun peek(offset: Int): T {
        return backingList[backingList.size - offset - 1]
    }

    override fun push(value: T) {
        backingList.add(value)
    }

    override fun pop(): T {
        return backingList.removeAt(backingList.size - 1)
    }

    override fun size(): Int {
        return backingList.size
    }
}
